extension/oauth2clientauthextension/extension.go

```diff
@@ -38,6 +39,14 @@ type ClientCredentialsAuthenticator struct {
 // ClientCredentialsAuthenticator implements ClientAuthenticator
 var _ configauth.ClientAuthenticator = (*ClientCredentialsAuthenticator)(nil)
 
+type errorWrappingTokenSource struct {
```

Comment:
We should be using the same token source inside perRPCCredentials as well. Also please rename the typr to something generic like tokenSourceDecorator once you remove the wrapper (error format string) from the type definition to module level error type (see my comments below)